diff --git a/src/openms/thirdparty/percolator/SanityCheck.cpp b/src/openms/thirdparty/percolator/SanityCheck.cpp
index 979ddfadbb..224dcb5d7b 100644
--- a/src/openms/thirdparty/percolator/SanityCheck.cpp
+++ b/src/openms/thirdparty/percolator/SanityCheck.cpp
@@ -24,8 +24,6 @@ using namespace std;
 #include "Normalizer.h"
 #include "SanityCheck.h"
 #include "Globals.h"
-#include "SqtSanityCheck.h"
-
 // Class doing the sanity check for non SQT file condition
 // In most sence a place holder as very little logic is build-in in this case
 
@@ -50,8 +48,6 @@ vector<double> SanityCheck::default_weights = vector<double>();
 SanityCheck* SanityCheck::initialize(string otherCall){
   if(initWeightFN != "" || initDefaultDirName.size() > 0) {
     return new SanityCheck();
-  } else if (otherCall.find(SqtSanityCheck::fingerPrint)!= string::npos){
-    return new SqtSanityCheck();
   } else {
     return new SanityCheck();
   }
