diff --git a/src/openms/thirdparty/percolator/PosteriorEstimator.cpp b/src/openms/thirdparty/percolator/PosteriorEstimator.cpp
index 8df2850cc4..1267f74672 100644
--- a/src/openms/thirdparty/percolator/PosteriorEstimator.cpp
+++ b/src/openms/thirdparty/percolator/PosteriorEstimator.cpp
@@ -33,8 +33,6 @@ using namespace std;
 #include "config.h"
 #endif
 
-#include "Option.h"
-#include "Version.h"
 #include "PosteriorEstimator.h"
 #include "Transform.h"
 #include "Globals.h"
@@ -604,106 +602,5 @@ string PosteriorEstimator::greeter() {
   return oss.str();
 }
 
-bool PosteriorEstimator::parseOptions(int argc, char** argv) {
-  // init
-  ostringstream intro;
-  intro << greeter() << endl;
-  intro << "Usage:" << endl;
-  intro << "   qvality [options] target_file null_file" << endl << "or"
-      << endl;
-  intro << "   qvality [options] pvalue_file" << endl << endl;
-  intro
-      << "target_file and null_file are files containing scores from a mixed model"
-      << endl;
-  intro
-      << "and a null model, each score separated with whitespace or line feed."
-      << endl;
-  intro
-      << "Alternatively, accuate p-value could be provided in a single file pvalue_file."
-      << endl << endl;
-  intro << "The method reports score, PEP, and q-value calculated directly from scores" << endl; 
-  intro << "Convergence of the method can be checked by comparing the average PEP of identifications" << endl;
-  intro << "above threshold with the q-value" << endl; 
-  CommandLineParser cmd(intro.str());
-  // finally parse and handle return codes (display help etc...)
-  cmd.defineOption("v",
-                   "verbose",
-                   "Set verbosity of output: 0=no processing info, 5=all, default is 2",
-                   "level");
-  cmd.defineOption("s",
-                   "epsilon-step",
-                   "The relative step size used as treshhold before cross validation error is calculated",
-                   "value");
-  cmd.defineOption("n",
-                   "number-of-bins",
-                   "The number of spline knots used when interpolating spline function. Default is 500.",
-                   "bins");
-  cmd.defineOption("c",
-                   "epsilon-cross-validation",
-                   "The relative crossvalidation step size used as treshhold before ending the iterations",
-                   "value");
-  cmd.defineOption("r",
-                   "reverse",
-                   "Indicating that the scoring mechanism is reversed, i.e., that low scores are better than higher scores",
-                   "",
-                   TRUE_IF_SET);
-  cmd.defineOption("o",
-                   "output-file",
-                   "Output results to file instead of stdout",
-                   "file");
-  cmd.defineOption("Y",
-                   "tdc-input",
-                   "Turns off the pi0 correction for search results from a concatenated database.",
-                   "",
-                   TRUE_IF_SET);
-  cmd.defineOption("d",
-                   "include-negative",
-                   "Include negative hits (decoy) probabilities in the results",
-		    "",
-		    TRUE_IF_SET);
-
-  cmd.parseArgs(argc, argv);
-  if (cmd.isOptionSet("verbose")) {
-    Globals::getInstance()->setVerbose(cmd.getInt("verbose", 0, 10));
-  }
-  if (cmd.isOptionSet("number-of-bins")) {
-    noIntervals = cmd.getInt("number-of-bins", 1, INT_MAX);
-  }
-  if (cmd.isOptionSet("epsilon-cross-validation")) {
-    BaseSpline::convergeEpsilon = cmd.getDouble("epsilon-cross-validation", 0.0, 1.0);
-  }
-  if (cmd.isOptionSet("epsilon-step")) {
-    BaseSpline::stepEpsilon = cmd.getDouble("epsilon-step", 0.0, 1.0);
-  }
-  if (cmd.isOptionSet("output-file")) {
-    resultFileName = cmd.options["output-file"];
-  }
-  if (cmd.isOptionSet("reverse")) {
-    PosteriorEstimator::setReversed(true);
-  }
-  if (cmd.isOptionSet("tdc-input")) {
-    PosteriorEstimator::setUsePi0(false);
-  }
-  if (cmd.isOptionSet("include-negative")) {
-    PosteriorEstimator::setNegative(true);
-  }
-  if (cmd.arguments.size() > 2) {
-    cerr << "Too many arguments given" << endl;
-    cmd.help();
-  }
-  if (cmd.arguments.size() == 0) {
-    cerr << "No arguments given" << endl;
-    cmd.help();
-  }
-  targetFile = cmd.arguments[0];
-  if (cmd.arguments.size() == 2) {
-    decoyFile = cmd.arguments[1];
-  } else {
-    PosteriorEstimator::setReversed(true);
-    pvalInput = true;
-  }
-  return true;
-}
-
 
 }}}  // namespace OpenMS::Internal::Percolator
diff --git a/src/openms/thirdparty/percolator/PosteriorEstimator.h b/src/openms/thirdparty/percolator/PosteriorEstimator.h
index 7be65b4716..84079d93fc 100644
--- a/src/openms/thirdparty/percolator/PosteriorEstimator.h
+++ b/src/openms/thirdparty/percolator/PosteriorEstimator.h
@@ -32,7 +32,6 @@ class PosteriorEstimator {
  public:
   PosteriorEstimator(){};
   virtual ~PosteriorEstimator(){};
-  bool parseOptions(int argc, char** argv);
   string greeter();
   int run();
   static void estimatePEP(std::vector<std::pair<double, bool> >& combined,
