--- msvc.py	2020-08-24 17:38:24.000000000 +0200
+++ msvc.py	2020-08-25 11:28:25.000000000 +0200
@@ -276,7 +276,7 @@
         ).decode('utf-16le', errors='replace')
     except subprocess.CalledProcessError as exc:
         raise distutils.errors.DistutilsPlatformError(
-            "Error executing {}".format(exc.cmd)
+            "Error executing {}. Error was {}".format(exc.cmd, exc.output)
         ) from exc
 
     env = {
